### PR TITLE
Check evalf

### PR DESCRIPTION
**Pull request by soraros**

### CODE REVIEW

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -168,6 +170,7 @@ jobs:
       - name: Run unit tests
         env:
           _image: ${{ steps.build.outputs.id }}
+          NUTILS_DEBUG: all
```
Comment (joostvanzwieten) [inline .github/workflows/test.yaml:173]: I don't think this works. You can set environment variables inside the container with `podman run ... --env ...`:

```sh
podman run --pull=never --rm -v "$PWD/tests:/app/tests:ro" -v "$PWD/examples:/app/examples:ro" --env NUTILS_DEBUG=all "$_image" -m unittest -bq
```

File: nutils/evaluable.py [hunk 2]
```diff
@@ -757,15 +758,34 @@ def _assparse(self):
...
+    def __get__(self, instance, owner):
+      evalf = self.evalf_obj(instance, owner)
+      @functools.wraps(evalf)
+      def evalf_with_check(*args, **kwargs):
+        res = evalf(*args, **kwargs)
+        if hasattr(instance, 'dtype'):
+          assert asdtype(res.dtype) == instance.dtype, ((instance.dtype, res.dtype), instance, res)
+          if hasattr(instance, 'ndim'):
+            assert res.ndim == instance.ndim
+          assert all(m == n for m, n in zip(res.shape, instance.shape) if isinstance(n, int)), 'shape mismatch'
```
Comment (joostvanzwieten) [inline nutils/evaluable.py:780]: If there is no `ndim`, then there should not be a `shape` either. Is the `if hasattr(instance, 'ndim')` test necessary?

I think not having a `dtype` doesn't imply there isn't a shape. Maybe we should place the `ndim` and `shape` tests under a separate `if hasattr(instance, 'shape')`?

Comment (joostvanzwieten) [reply nutils/evaluable.py:780]: What do you think of the following:

```python
assert not hasattr(instance, 'dtype') or asdtype(res.dtype) == instance.dtype, ((instance.dtype, res.dtype), instance, res)
assert not hasattr(instance, 'ndim') or res.ndim == instance.ndim
assert not hasattr(instance, 'shape') or all(m == n for m, n in zip(res.shape, instance.shape) if isinstance(n, int)), 'shape mismatch'
```

**Comment by joostvanzwieten** [comment]
Can you reorder the commits such that all of them pass the unittests? Fixes first, then adding the checker.

File: nutils/debug_flags.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+_flags = _all if _env == 'all' \
+    else set(filter(None, _env.lower().split(':'))) | _default if __debug__ \
+    else set(filter(None, _env.lower().split(':')))
+assert _flags <= _all, 'Unknown debug flags: {}'.format(_flags - _all)
+
+for f in _all:
+  locals()[f] = f in _flags
+
+if __name__ == '__main__':
+  print('Debug flags set:', ', '.join(_flags))
```
Comment (gertjanvanzwieten) [inline nutils/debug_flags.py:37]: The following code is a bit easier on the eyes:

```
import os, warnings

_flags = dict.fromkeys(os.getenv('NUTILS_DEBUG', '').lower().split(':'), True)
_all = _flags.pop('all', False)

sparse = _flags.pop('sparse', _all  or __debug__) # check sparse chunks in evaluable
lower = _flags.pop('lower', _all or __debug__) # check lowered shape, dtype in function
evalf = _flags.pop('evalf', _all) # check evaluated arrays in evaluable

if _flags:
  warnings.warn('unused debug flags: {}'.format(', '.join(_flags)))
```

Comment (soraros) [reply nutils/debug_flags.py:37]: Done, thanks!

File: .github/workflows/test.yaml (hunk 1 above)
Comment (soraros) [reply .github/workflows/test.yaml]: Done. Thanks!

File: nutils/evaluable.py (hunk 2 above)
Comment (soraros) [reply nutils/evaluable.py]: Done. Thanks!

**Review by joostvanzwieten — APPROVED** [review]

File: nutils/debug_flags.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import os, warnings
+
+_env = dict.fromkeys(os.getenv('NUTILS_DEBUG', '').lower().split(':'), True)
```
Comment (gertjanvanzwieten) [inline nutils/debug_flags.py:23]: Just noticed we need to filter empty strings; currently the warning is always triggered.
```
_env = dict.fromkeys(filter(None, os.getenv('NUTILS_DEBUG', '').lower().split(':')), True)
```

**Review by gertjanvanzwieten — CHANGES_REQUESTED** [review]
Could you add the filter and rebase? Then I'll merge it as soon as the tests pass again.

File: nutils/debug_flags.py (hunk 4 above)
Comment (soraros) [reply nutils/debug_flags.py]: Done. Thanks!

**Review by gertjanvanzwieten — APPROVED** [review]

**Pull request merged**